zebra/zebra_rnh.c

```diff
@@ -623,10 +628,34 @@ zebra_rnh_resolve_nexthop_entry(struct zebra_vrf *zvrf, afi_t afi,
 				break;
 		}
 
-		/* Route entry found, we're done; else, walk up the tree. */
+		/* Route entry found, we're done; else, walk up the tree.
+		 *
+		 * BGP routes are usually recursive, though. Keep resolving
+		 * on the nexthop until hitting a non-BGP route.
+		 */
 		if (re) {
-			*prn = rn;
-			return re;
+			if (re->type != ZEBRA_ROUTE_BGP) {
```

Comment:
I understand what you are attempting to do now.  Please remove the test for the re->type and use the re->flags to see if the route has been set with ZEBRA_FLAG_ALLOW_RECURSION.  Recursion is what you is really interesting here *not* bgp routes.  

Additionally please modify the test to actually test the `show ip nht json` output.

Reply:
gah my english.  `Recursion is what is interesting here, *not* bgp routes`

Reply:
Thanks for looking into the issue. Yes, it is about recursion. I was trying to limit the scope, and not dealing with the recursive static routes (which is not a good practice IMO).

Let us think about whether it's ok to limit the fix to bgp routes for now.

I will add include "show ip nht json" output.

Reply:
Code generalized for recursive nexthop resolution.